pyflow/RCDT2/Nodes/core.py

```diff
@@ -0,0 +1,183 @@
+        super().__init__(name)
+        self.service_type: type
+
+        self.createInputPin("In", "ExecPin", callback=self.execute_parallel)
+        self.exec_out: PinBase = self.createOutputPin("Out", "ExecPin")
+
+        for service_part in ["Request", "Response"]:
+            self.create_data_pins(service_part)
+
+        self.topic_name = underscore(self.service_type.__name__)
+        self.client = PyflowRosBridge.node.create_client(
+            self.service_type, self.topic_name
+        )
+
+    def execute(self, *_args: any, **_kwargs: any) -> None:
+        request = self.create_request_from_pins()
+        response = self.call_service(request)
+        if not response.success:
+            logger.error(
+                f"Call to {self.topic_name} with type {self.service_type.__name__} was unsuccessfull."
+            )
+            return
+        logger.info("Response was succesfull.")
+        self.set_response_to_pins(response)
+        self.exec_out.call()
+
+    def create_data_pins(self, service_part: str) -> None:
+        msg: type = getattr(self.service_type, service_part)
+        fields_and_field_types: dict[str, str] = msg.get_fields_and_field_types()
+        for field, field_type in fields_and_field_types.items():
+            pin_type = get_pin_type(field_type)
+            if msg.__name__ == self.service_type.__name__ + "_Request":
+                pin = self.createInputPin(field, pin_type)
+                self.pin_manager.add_input_pin(field, pin)
+            elif msg.__name__ == self.service_type.__name__ + "_Response":
+                pin = self.createOutputPin(field, pin_type)
+                self.pin_manager.add_output_pin(field, pin)
+            else:
+                logger.warning("Message is not a Request or a Response. Skip.")
+
+    def create_request_from_pins(self) -> object:
+        request = self.service_type.Request()
+        for pin_name in self.pin_manager.get_input_pin_names():
+            data = self.pin_manager.get_data(pin_name)
+            setattr(request, pin_name, data)
+        return request
+
+    def set_response_to_pins(self, response: object) -> None:
+        for pin_name in self.pin_manager.get_output_pin_names():
+            data = getattr(response, pin_name)
+            self.pin_manager.set_data(pin_name, data)
+
+    def call_service(self, request: object) -> object:
+        logger.info("Calling service...")
+        response = self.service_type.Response()
+        response.success = False
+        if not self.client.wait_for_service(3):
+            logger.error("Service not available. Exit.")
+            return response
+
+        future = self.client.call_async(request)
+        timeout = 20
+        waited = 0
+        while not future.done():
+            if waited >= timeout:
+                future.cancel()
+                logger.error("No response from service. Exit.")
+                return response
+            sleep(1)
+            waited += 1
```

Comment:
Je moet nu zelf bijhouden hoe lang je gewacht hebt. Een typisch patroon in Python is while-else. In de while loop komt de check of de timeout al om is. In de while is een check op het criterium dat je wil voldoen (in ons geval future.done()) met een break. Als de future op tijd terug komt, dan wordt de else-clause niet geactiveerd. Als we de timeout bereiken dan wel.

PS: waarschijnlijk gebruik je liever `from time import sleep, time`, en hier `time()` ipv `time.time()`

```suggestion
        start_time = time.time()
        timeout = 20
        while time.time() - start_time < 20:
            if future.done():
                break
            sleep(1)
        else:
            future.cancel()
            logger.error("No response from service. Exit.")
            return response
```